data/gcalctool.1
.\" Copyright (c) 1987-2008 - Sun Microsystems, Inc.
.TH GCALCTOOL 1 "17 June 2009"
.SH NAME
gcalctool \- a desktop calculator
.SH SYNOPSIS
.B gcalctool
[
.I OPTION
] ...
.SH DESCRIPTION
.B gcalctool
is the official calculator of the GNOME desktop environment.
.SH OPTIONS
.TP
.B \-s, \-\-solve
Solve the equation provided following this option.
.TP
.B \-u, \-\-unittest
Run the unit tests.
.TP
.B \-v, \-\-version
Output version information and exit.
.TP
.B \-h, \-?, \-\-help
Prints the command line options.
.SH SEE ALSO
.PD 0
.TP 18
.B gcalccmd
.sp
.LP


NEWS
Overview of changes in gcalctool 6.0.0

    * Use /org/gnome as location for gsettings keys.  This will mean
      gcalctool 5.31.x users will have their settings reset to defaults.
    * Correctly choose the first conversion when none/invalid defined
      (Bug #645663)

Overview of changes in gcalctool 5.91.93

    * Translation updates

Overview of changes in gcalctool 5.91.92

    * Add gcalccmd man page (Jean Schurger, Bug #645138)
    * Remove references to ~/.gcalctoolrc (Jean Schurger, Bug #645137)

Overview of changes in gcalctool 5.91.91

    * Disable localized digits (Bug #644980)
    * Fix crash doing bitwise operations with large numbers (Bug #637328)
    * Don't use automatic exponential format when not in base 10
    * Only use thousands separators in base 10
    * Set currency on first start (Bug #640818)
    * Support entering super/subscript with numeric keypad (Bug #634904)

Overview of changes in gcalctool 5.91.90

    * Integrate unittests into build process

Overview of changes in gcalctool 5.91.90

    * Fix grammar mistakes
    * Fix translated conversions (Bug #641166)

Overview of changes in gcalctool 5.91.6

    * Add fixed rate of 15.6466 EEK = 1 EUR (Bug #639332)
    * Use full name of currency in conversion bar
    * Add temperature conversion
    * Update conversion bar when currencies are updated
    * Fix incorrect display of complex numbers
    * Now able to change currencies from advanced mode

Overview of changes in gcalctool 5.91.5

    * Make new variable buttons that combines add/set/get and clear into one
      dialog.
    * Add a decimal point button to the programming mode
    * Make the numeric keypad produce numbers even when numlock is on

Overview of changes in gcalctool 5.91.4

    * Use new Indian Rupee sign ₹ (Bug #636587, Carlos Cejudo)
    * Use currency rates from the IMF as well as the ECB to add 24 new rates:
      AED, BHD, BND, BWP, CLP, COP, DZD, ILS, IRR, KWD, KZT, LKR,
      LYD, MUR, NPR, OMR, PEN, PKR, QAR, SAR, TND, TTD, UYU, VEF
    * Don't hide GCalctool in KDE
    * Convert numeric keypad separator into locale specific version
      (Bug #632794)

Overview of changes in gcalctool 5.91.3

    * Fix incorrect calculation of tanh
    * Fix dropping of decimal points when thousand separator is '.' (Bug #635517)
    * Improve conversion bar
    * Fix crash in scientific mode
    * Add new automatic display mode that switches from fixed to scientific when
      numbers are too large

Overview of changes in gcalctool 5.91.2

    * Fix broken thousands separators code (Bug #628908)
    * Don't make window resizable, it is not supposed to be
    * Use icons for undo and clear buttons as the translations cause them to be
      oversized in some languages
    * Handle Han digits (〇〡〢〣〤〥〦〧〨〩)
    * If a thousands separator is not defined then use ' ' (Bug #632182)
    * Fix incorrect error text for acosh function (Bug #632614)

Overview of changes in gcalctool 5.91.1

    * Clear display when an answer is displayed and a numeric point is inserted
    * Add GUI for unit conversations (bug #539830)
    * Fix the definition of ounces
    * Remove trailing whitespace when entering super-/subscript mode
      (bug #631988)
    * Calculate answers in separate thread (bug #576371)
    * Stop '*'+backspace+'*' being converted into '^'
    * Updated translations

Overview of changes in gcalctool 5.91.0

    * Port to GTK+-3 (bug #630115, William Jon McCann)
    * Don't crash on pasting empty clipboard
    * Mark more strings as translatable (bug #630899, Gabor Kelemen)
    * Make calculator window resizable (bug #630594)
    * Put back visible thousands separators (bug #628908)
    * Make thousands separators and radix points respect the users locale
    * Updated translations

Overview of changes in gcalctool 5.32.0

    * Updated translations

Overview of changes in gcalctool 5.31.91

    * Increase accuracy of internal calculations, don't crash printing numbers
      when they get too large.
    * Updated translations
 
Overview of changes in gcalctool 5.31.90

    * Updated translations

Overview of changes in gcalctool 5.31.6

    * Fix double pasting into display with middle button
    * Add a11y names to widgets

Overview of changes in gcalctool 5.31.5

    * Call gsettings_sync() before quit
 
Overview of changes in gcalctool 5.31.4

    * Show currency conversion below display

    * Use gsettings instead of gconf
 
Overview of changes in gcalctool 5.31.3

    * Support complex trigonometry

    * Add complex argument button to scientific mode
    
    * Make popup button for additional functions, with integer/fractional
      component, floor, ceiling, round and sign functions.

Overview of changes in gcalctool 5.31.2

    * Replace backspace button with undo button

    * Remove memory buttons from basic mode (too complex)

    * Support complex arithmetic

Overview of changes in gcalctool 5.31.1

    * Dynamically load button layouts to speed up startup time

    * Add colours to buttons (Bug #157962)

    * Merge advanced and scientific mode

    * Relayout buttons

    * Use arrows on superscript buttons (i.e. '↑n' instead of '⎕ⁿ')

    * Don't change window title to match mode (no need to)

    * Rename View menu to Mode

    * Make inverses of variables work (Bug #614979)

    * Show answer in bold (Bug #505168)

    * Reimplement undo and redo

    * Automatically add missing closing brackets (Bug #615243)

    * Show all defined variables in variable popups

Overview of changes in gcalctool 5.30.0

    * Fix errors in variable exponents (e.g. xy²)

    * Handle powers in convesion units, (e.g. 1m² in mm²)

    * Don't clear ans when clearing display (Bug #613552)
    
    * Handle equations like (x+3y)(2x-4y)

Overview of changes in gcalctool 5.29.92

    * Add shortcuts to base buttons (Ctrl+B, Ctrl+O, Ctrl+D, Ctrl+H)

    * Convert number in display if it has a base and a base button is pressed


src/gcalctool.c
/*
 * Copyright (C) 1987-2008 Sun Microsystems, Inc. All Rights Reserved.
 * Copyright (C) 2008-2011 Robert Ancell
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 2 of the License, or (at your option) any later
 * version. See http://www.gnu.org/copyleft/gpl.html the full text of the
 * license.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <locale.h>
#include <glib/gi18n.h>

#include "math-window.h"
#include "mp-equation.h"
#include "unit-manager.h"

static GSettings *settings = NULL;

static MathWindow *window;

static void
version(const gchar *progname)
{
    /* NOTE: Is not translated so can be easily parsed */
    fprintf(stderr, "%1$s %2$s\n", progname, VERSION);
}


static int
do_convert(const MPNumber *x, const char *x_units, const char *z_units, MPNumber *z, void *data)
{
    return unit_manager_convert_by_symbol(unit_manager_get_default(), x, x_units, z_units, z);
}


static void
solve(const char *equation)
{
    MPEquationOptions options;
    MPErrorCode error;
    MPNumber result;
    char *result_str;

    memset(&options, 0, sizeof(options));
    options.base = 10;
    options.wordlen = 32;
    options.angle_units = MP_DEGREES;
    options.convert = do_convert;

    error = mp_equation_parse(equation, &options, &result, NULL);
    if(error == PARSER_ERR_MP) {
        fprintf(stderr, "Error: %s\n", mp_get_error());
        exit(1);
    }
    else if(error != 0) {
        fprintf(stderr, "Error: %s\n", mp_error_code_to_string(error));
        exit(1);
    }
    else {
        result_str = mp_serializer_to_string(mp_serializer_new(MP_DISPLAY_FORMAT_AUTOMATIC, 10, 9), &result);
        printf("%s\n", result_str);
        exit(0);
    }
}


static void
usage(const gchar *progname, gboolean show_application, gboolean show_gtk)
{
    fprintf(stderr,
            /* Description on how to use gcalctool displayed on command-line */
            _("Usage:\n"
              "  %s — Perform mathematical calculations"), progname);

    fprintf(stderr,
            "\n\n");

    fprintf(stderr,
            /* Description on gcalctool command-line help options displayed on command-line */
            _("Help Options:\n"
              "  -v, --version                   Show release version\n"
              "  -h, -?, --help                  Show help options\n"
              "  --help-all                      Show all help options\n"
              "  --help-gtk                      Show GTK+ options"));
    fprintf(stderr,
            "\n\n");

    if (show_gtk) {
        fprintf(stderr,
                /* Description on gcalctool command-line GTK+ options displayed on command-line */
                _("GTK+ Options:\n"
                  "  --class=CLASS                   Program class as used by the window manager\n"
                  "  --name=NAME                     Program name as used by the window manager\n"
                  "  --screen=SCREEN                 X screen to use\n"
                  "  --sync                          Make X calls synchronous\n"
                  "  --gtk-module=MODULES            Load additional GTK+ modules\n"
                  "  --g-fatal-warnings              Make all warnings fatal"));
        fprintf(stderr,
                "\n\n");
    }

    if (show_application) {
        fprintf(stderr,
                /* Description on gcalctool application options displayed on command-line */
                _("Application Options:\n"
                  "  -s, --solve <equation>          Solve the given equation"));
        fprintf(stderr,
                "\n\n");
    }
}


static void
get_options(int argc, char *argv[])
{
    int i;
    char *progname, *arg;

    progname = g_path_get_basename(argv[0]);

    for (i = 1; i < argc; i++) {
        arg = argv[i];

        if (strcmp(arg, "-v") == 0 ||
            strcmp(arg, "--version") == 0) {
            version(progname);
            exit(0);
        }
        else if (strcmp(arg, "-h") == 0 ||
                 strcmp(arg, "-?") == 0 ||
                 strcmp(arg, "--help") == 0) {
            usage(progname, TRUE, FALSE);
            exit(0);
        }
        else if (strcmp(arg, "--help-all") == 0) {
            usage(progname, TRUE, TRUE);
            exit(0);
        }
        else if (strcmp(arg, "--help-gtk") == 0) {
            usage(progname, FALSE, TRUE);
            exit(0);
        }
        else if (strcmp(arg, "-s") == 0 ||
            strcmp(arg, "--solve") == 0) {
            i++;
            if (i >= argc) {
                fprintf(stderr,
                        /* Error printed to stderr when user uses --solve argument without an equation */
                        _("Argument --solve requires an equation to solve"));
                fprintf(stderr, "\n");
                exit(1);
            }
            else
                solve(argv[i]);
        }
        else {
            fprintf(stderr,
                    /* Error printed to stderr when user provides an unknown command-line argument */
                    _("Unknown argument '%s'"), arg);
            fprintf(stderr, "\n");
            usage(progname, TRUE, FALSE);
            exit(1);
        }
    }
}


static void
quit_cb(MathWindow *window)
{
    MathEquation *equation;
    MathButtons *buttons;

    equation = math_window_get_equation(window);
    buttons = math_window_get_buttons(window);

    g_settings_set_int(settings, "accuracy", math_equation_get_accuracy(equation));
    g_settings_set_int(settings, "word-size", math_equation_get_word_size(equation));
    g_settings_set_int(settings, "base", math_buttons_get_programming_base(buttons));
    g_settings_set_boolean(settings, "show-thousands", math_equation_get_show_thousands_separators(equation));
    g_settings_set_boolean(settings, "show-zeroes", math_equation_get_show_trailing_zeroes(equation));
    g_settings_set_enum(settings, "number-format", math_equation_get_number_format(equation));
    g_settings_set_enum(settings, "angle-units", math_equation_get_angle_units(equation));
    g_settings_set_enum(settings, "button-mode", math_buttons_get_mode(buttons));
    g_settings_set_string(settings, "source-currency", math_equation_get_source_currency(equation));
    g_settings_set_string(settings, "target-currency", math_equation_get_target_currency(equation));
    g_settings_set_string(settings, "source-units", math_equation_get_source_units(equation));
    g_settings_set_string(settings, "target-units", math_equation_get_target_units(equation));
    g_settings_sync();

    gtk_main_quit();
}


int
main(int argc, char **argv)
{
    MathEquation *equation;
    int accuracy = 9, word_size = 64, base = 10;
    gboolean show_tsep = FALSE, show_zeroes = FALSE;
    MpDisplayFormat number_format;
    MPAngleUnit angle_units;
    ButtonMode button_mode;
    gchar *source_currency, *target_currency;
    gchar *source_units, *target_units;

    g_type_init();

    setlocale(LC_ALL, "");
    bindtextdomain(GETTEXT_PACKAGE, LOCALE_DIR);
    bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
    textdomain(GETTEXT_PACKAGE);

    /* Seed random number generator. */
    srand48((long) time((time_t *) 0));

    get_options(argc, argv);

    settings = g_settings_new ("org.gnome.gcalctool");
    accuracy = g_settings_get_int(settings, "accuracy");
    word_size = g_settings_get_int(settings, "word-size");
    base = g_settings_get_int(settings, "base");
    show_tsep = g_settings_get_boolean(settings, "show-thousands");
    show_zeroes = g_settings_get_boolean(settings, "show-zeroes");
    number_format = g_settings_get_enum(settings, "number-format");
    angle_units = g_settings_get_enum(settings, "angle-units");
    button_mode = g_settings_get_enum(settings, "button-mode");
    source_currency = g_settings_get_string(settings, "source-currency");
    target_currency = g_settings_get_string(settings, "target-currency");
    source_units = g_settings_get_string(settings, "source-units");
    target_units = g_settings_get_string(settings, "target-units");

    equation = math_equation_new();
    math_equation_set_accuracy(equation, accuracy);
    math_equation_set_word_size(equation, word_size);
    math_equation_set_show_thousands_separators(equation, show_tsep);
    math_equation_set_show_trailing_zeroes(equation, show_zeroes);
    math_equation_set_number_format(equation, number_format);
    math_equation_set_angle_units(equation, angle_units);
    math_equation_set_source_currency(equation, source_currency);
    math_equation_set_target_currency(equation, target_currency);
    math_equation_set_source_units(equation, source_units);
    math_equation_set_target_units(equation, target_units);
    g_free(source_currency);
    g_free(target_currency);
    g_free(source_units);
    g_free(target_units);

    gtk_init(&argc, &argv);

    window = math_window_new(equation);
    g_signal_connect(G_OBJECT(window), "quit", G_CALLBACK(quit_cb), NULL);
    math_buttons_set_programming_base(math_window_get_buttons(window), base);
    math_buttons_set_mode(math_window_get_buttons(window), button_mode); // FIXME: We load the basic buttons even if we immediately switch to the next type

    gtk_widget_show(GTK_WIDGET(window));
    gtk_main();

    return(0);
}
